daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/SpecifiedLengthParsers.scala

```diff
@@ -85,7 +85,9 @@ sealed abstract class SpecifiedLengthParserBase(eParser: Parser, erd: RuntimeDat
     if (pState.processorStatus ne Success) return
     val finalEndPos0b = startingBitPos0b + nBits
 
-    captureValueLength(pState, ULong(startingBitPos0b), ULong(dis.bitPos0b))
+    // if we haven't already set the value length, set it now
+    if (pState.infoset.valueLength.isEndUndef)
+      captureValueLength(pState, ULong(startingBitPos0b), ULong(dis.bitPos0b))
```

Comment:
Is the issue here that some `eParsers` do not capture their own length and expect it to be captured here, but other parsers do capture value length (because they handle things like trimming), so sometimes there are two parsers trying to capture length for the same thing?

That makes me wonder if we shouldn't be capturing value length here at all, and instead we should be using the `CaptureValueLengthStart/End` parsers for those parsers that don't capture value length?

ElementBaseRuntime1Mixin has a `capturedByParsers` val that tries to list which parsers capture their own value length. Maybe that logic isn't completely correct, and the work around (that feels wrong) is we just always capture the length here?

Reply:
> Is the issue here that some `eParsers` do not capture their own length and expect it to be captured here, but other parsers do capture value length (because they handle things like trimming), so sometimes there are two parsers trying to capture length for the same thing?

That's right. For example, SpecifiedLengthExplicitParser doesn't capture its own length.

> That makes me wonder if we shouldn't be capturing value length here at all, and instead we should be using the CaptureValueLengthStart/End parsers for those parsers that don't capture value length?

I'll look into implementing this

Reply:
> ElementBaseRuntime1Mixin has a capturedByParsers val that tries to list which parsers capture their own value length. Maybe that logic isn't completely correct, and the work around (that feels wrong) is we just always capture the length here?

I think it's the "isReferenced" boolean that may be the issue, if we remove that, then we'll always captureValueLength if capturedByParsers is false

Reply:
The `isReferenced` variable is used to avoid capturing value lengths that will never be used. For example, if we have something like this:

```xml
<element name="a" type="xs:int" ... />
```
If nothing ever references the value length of `a, then we shouldn't insert `CaptureValueLengthStart/End` parsers, regardless of the value of `capturedByParsers`. But if something does reference the length, e.g.
```xml
<element name="len" type="xs:int" dfdl:inputValueCalc="{ dfdl:valueLength(../a, 'bytes')) } />
```
The we do need to capture the value length of `a`. So `isReferenced` is needed to disabled capturing when it's not needed. I don't think we can get rid of it.

Usually we capture the length via the `CaptureValueLengthStart/End` parsers, but sometimes those parsers aren't smart enough so it must be done by a separate parser that has enough information to capture length correctly (e.g. StringOfSpecifiedLength that knows about pad/trim stuff and how to exclude it from the value length). The `capturedByParsers` val is supposed to figure out which are those special cases so we don't unnecessarily insert `CaptureValueLengthStart/End` parsers for parsers that do it themselves.

It seems like SpecifiedLengthParserBase doesn't have enough information to know how to correctly capture value length (since sometimes the subparser is the thing that knows how to do it). So that probably means it shouldn't be capturing the length at all. But that then means other parsers that relied on that behavior either need to call `captureValueLength` themselves, or `capturedByParsers` needs to be false in some cases so that we insert the `CaptureValueLengthStart/End` parsers.

Reply:
All the complexity around whether to insert a value-length capture parser at the start/end is about symmetry of the parser with the way the unparser works. 

Capturing length when parsing is easy. Record start position when parser is entered, ending position when it is exited.  Probably so light weight that it's not worth trying to optimize it away.

Unparsing... not so much. The unparser has to unwind and allow unparsing to continue while the value length is suspended. Definitely we want none of this overhead for elements which no expression references their length. We only want capture of length overheads when it is essential. I am also hoping that some of this is simpler for prefixed lengths, which don't work for unparsing with suspensions, etc. currently.